File: backend/app/routers/drops.py
from datetime import datetime, timezone
from typing import Annotated
from uuid import UUID, uuid4

from app.core.security import (
    decrypt_share_token,
    encrypt_share_token,
    generate_share_token,
    hash_share_token,
)
from app.core.storage_limits import (
    PLATFORM_MAX_STORAGE,
    USER_MAX_ACTIVE_STORAGE,
    USER_MAX_DROP_STORAGE,
    USER_MAX_FILE_SIZE,
    USER_MAX_FILES_PER_DROP,
)
from app.db.database import DbSession
from app.dependencies.auth import CurrentUser
from app.models.drop import Drop
from app.models.drop_file import DropFile
from app.models.user import User
from app.schemas.drop import (
    DropCreate,
    DropCreateResponse,
    DropListResponse,
    DropOut,
    DropStatus,
    DropUpdate,
    ShareTokenResponse,
)
from app.schemas.drop_file import DropFileOut, FileUploadRequest, FileUploadResponse
from app.services.storage import (
    create_presigned_upload,
    delete_file,
    get_file_metadata,
    promote_file,
)
from app.services.storage_usage import get_active_user_storage
from fastapi import APIRouter, HTTPException, Query, status
from sqlalchemy import func, or_, select

router = APIRouter(
    prefix="/drops",
    tags=["Drops"],
)


@router.get(
    "",
    response_model=DropListResponse,
)
def get_my_drops(
    current_user: CurrentUser,
    db: DbSession,
    page: Annotated[int, Query(ge=1)] = 1,
    page_size: Annotated[int, Query(ge=1, le=100)] = 10,
    search: Annotated[str | None, Query(max_length=100)] = None,
    drop_status: Annotated[
        DropStatus,
        Query(alias="status"),
    ] = DropStatus.ACTIVE,
):
    now = datetime.now(timezone.utc)

    query = select(Drop).where(
        Drop.owner_id == current_user.id,
    )

    if drop_status == DropStatus.REVOKED:
        query = query.where(
            Drop.revoked_at.is_not(None),
        )
    elif drop_status == DropStatus.EXPIRED:
        query = query.where(
            Drop.revoked_at.is_(None),
            Drop.expires_at <= now,
        )
    elif drop_status == DropStatus.CONSUMED:
        query = query.where(
            Drop.revoked_at.is_(None),
            Drop.expires_at > now,
            Drop.max_views <= Drop.view_count,
        )
    else:  # Active
        query = query.where(
            Drop.revoked_at.is_(None),
            Drop.expires_at > now,
            Drop.max_views > Drop.view_count,
        )

    if search:
        query = query.where(
            or_(
                Drop.title.ilike(f"%{search}%"),
                Drop.content.ilike(f"%{search}%"),
            )
        )

    count_query = select(func.count()).select_from(query.subquery())
    total = db.scalar(count_query) or 0

    offset = (page - 1) * page_size

    query = query.order_by(Drop.created_at.desc()).offset(offset).limit(page_size)

    drops = db.scalars(query).all()

    return DropListResponse(
        items=drops,
        total=total,
        page=page,
        page_size=page_size,
    )


@router.post(
    "",
    response_model=DropCreateResponse,
    status_code=status.HTTP_201_CREATED,
)
def create_drop(
    data: DropCreate,
    current_user: CurrentUser,
    db: DbSession,
):
    share_token = generate_share_token()

    drop = Drop(
        owner_id=current_user.id,
        title=data.title,
        content=data.content,
        token_hash=hash_share_token(share_token),
        encrypted_token=encrypt_share_token(share_token),
        expires_at=data.expires_at,
        max_views=data.max_views,
    )

    db.add(drop)

    try:
        db.commit()
    except Exception:
        db.rollback()
        raise

    db.refresh(drop)

    return DropCreateResponse(
        **DropOut.model_validate(drop).model_dump(),
        share_token=share_token,
    )


@router.get(
    "/{drop_id}/share-token",
    response_model=ShareTokenResponse,
)
def get_share_token(
    drop_id: UUID,
    current_user: CurrentUser,
    db: DbSession,
):
    drop = db.scalar(
        select(Drop).where(
            Drop.id == drop_id,
            Drop.owner_id == current_user.id,
        )
    )

    if drop is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Drop not found",
        )

    if drop.encrypted_token is None:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail="Share token is not available for this Drop",
        )

    share_token = decrypt_share_token(drop.encrypted_token)

    return ShareTokenResponse(
        share_token=share_token,
    )


@router.get(
    "/{drop_id}",
    response_model=DropOut,
)
def get_my_drop(
    drop_id: UUID,
    current_user: CurrentUser,
    db: DbSession,
):
    drop = db.scalar(
        select(Drop).where(
            Drop.id == drop_id,
            Drop.owner_id == current_user.id,
        )
    )

    if drop is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Drop not found",
        )

    return drop


@router.get(
    "/{drop_id}/files",
    response_model=list[DropFileOut],
)
def get_my_drop_files(
    drop_id: UUID,
    current_user: CurrentUser,
    db: DbSession,
):
    drop = db.scalar(
        select(Drop).where(
            Drop.id == drop_id,
            Drop.owner_id == current_user.id,
        )
    )

    if drop is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Drop not found",
        )

    return db.scalars(
        select(DropFile)
        .where(
            DropFile.drop_id == drop.id,
            DropFile.uploaded_at.is_not(None),
            DropFile.storage_deleted_at.is_(None),
        )
        .order_by(DropFile.created_at.asc(), DropFile.id.asc())
    ).all()


@router.patch(
    "/{drop_id}",
    response_model=DropOut,
)
def update_my_drop(
    drop_id: UUID,
    data: DropUpdate,
    current_user: CurrentUser,
    db: DbSession,
):
    drop = db.scalar(
        select(Drop).where(
            Drop.id == drop_id,
            Drop.owner_id == current_user.id,
        )
    )

    if drop is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Drop not found",
        )

    if drop.revoked_at is not None:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail="A revoked Drop cannot be updated",
        )

    now = datetime.now(timezone.utc)

    if drop.expires_at <= now:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail="An expired Drop cannot be updated",
        )

    update_data = data.model_dump(exclude_unset=True)

    if "max_views" in update_data and update_data["max_views"] < drop.view_count:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Max views cannot be lower than the current view count",
        )

    for field, value in update_data.items():
        setattr(drop, field, value)

    try:
        db.commit()
    except Exception:
        db.rollback()
        raise

    db.refresh(drop)

    return drop


@router.delete(
    "/{drop_id}",
    status_code=status.HTTP_204_NO_CONTENT,
)
def revoke_drop(
    drop_id: UUID,
    current_user: CurrentUser,
    db: DbSession,
):
    drop = db.scalar(
        select(Drop).where(
            Drop.id == drop_id,
            Drop.owner_id == current_user.id,
        )
    )

    if drop is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Drop not found",
        )

    if drop.revoked_at is None:
        drop.revoked_at = datetime.now(timezone.utc)

        try:
            db.commit()
        except Exception:
            db.rollback()
            raise


@router.post(
    "/{drop_id}/files/presign",
    response_model=FileUploadResponse,
    status_code=status.HTTP_201_CREATED,
)
def create_file_upload(
    drop_id: UUID,
    payload: FileUploadRequest,
    db: DbSession,
    current_user: CurrentUser,
):
    now = datetime.now(timezone.utc)

    # Serialize upload reservations for this user.
    db.execute(
        select(User).where(User.id == current_user.id).with_for_update()
    ).scalar_one()

    drop = db.scalar(
        select(Drop).where(
            Drop.id == drop_id,
            Drop.owner_id == current_user.id,
        )
    )

    if drop is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Drop not found",
        )

    # File uploads only belong to active Drops.
    if (
        drop.revoked_at is not None
        or drop.expires_at <= now
        or drop.view_count >= drop.max_views
    ):
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail="Files cannot be added to an inactive Drop",
        )

    # Individual file limit: 10 MiB.
    if payload.size_bytes > USER_MAX_FILE_SIZE:
        raise HTTPException(
            status_code=status.HTTP_413_CONTENT_TOO_LARGE,
            detail="File exceeds the 10 MB limit",
        )

    # Pending uploads count too.
    existing_files = (
        db.scalar(
            select(func.count(DropFile.id)).where(
                DropFile.drop_id == drop.id,
                DropFile.storage_deleted_at.is_(None),
            )
        )
        or 0
    )

    if existing_files >= USER_MAX_FILES_PER_DROP:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail="This Drop already has the maximum number of files",
        )

    drop_storage = (
        db.scalar(
            select(func.coalesce(func.sum(DropFile.size_bytes), 0)).where(
                DropFile.drop_id == drop.id,
                DropFile.storage_deleted_at.is_(None),
            )
        )
        or 0
    )

    if drop_storage + payload.size_bytes > USER_MAX_DROP_STORAGE:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail="This Drop would exceed its storage limit",
        )

    active_user_storage = get_active_user_storage(db, current_user.id, now)

    if active_user_storage + payload.size_bytes > USER_MAX_ACTIVE_STORAGE:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail="Your active storage quota would be exceeded",
        )

    platform_storage = (
        db.scalar(
            select(func.coalesce(func.sum(DropFile.size_bytes), 0)).where(
                DropFile.storage_deleted_at.is_(None),
            )
        )
        or 0
    )

    if platform_storage + payload.size_bytes > PLATFORM_MAX_STORAGE:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="SafeDrop storage capacity has been reached",
        )

    file_id = uuid4()

    temporary_key = f"pending/{drop.id}/{file_id}"

    drop_file = DropFile(
        id=file_id,
        drop_id=drop.id,
        original_name=payload.original_name,
        storage_key=temporary_key,
        content_type=payload.content_type,
        size_bytes=payload.size_bytes,
    )

    db.add(drop_file)
    db.commit()

    presigned = create_presigned_upload(
        storage_key=temporary_key,
        content_type=payload.content_type,
        max_size=payload.size_bytes,
        expires_in=300,
    )

    return FileUploadResponse(
        file_id=file_id,
        upload_url=presigned["url"],
        fields=presigned["fields"],
        expires_in=300,
    )


@router.post(
    "/{drop_id}/files/{file_id}/complete",
    response_model=DropFileOut,
)
def complete_file_upload(
    drop_id: UUID,
    file_id: UUID,
    db: DbSession,
    current_user: CurrentUser,
):
    now = datetime.now(timezone.utc)

    # Make sure this is the user's Drop.
    drop = db.scalar(
        select(Drop).where(
            Drop.id == drop_id,
            Drop.owner_id == current_user.id,
        )
    )

    if drop is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Drop not found",
        )

    # Lock this file reservation so two completion requests
    # cannot finalize it simultaneously.
    drop_file = db.scalar(
        select(DropFile)
        .where(
            DropFile.id == file_id,
            DropFile.drop_id == drop.id,
        )
        .with_for_update()
    )

    if drop_file is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="File not found",
        )

    # Makes the endpoint safe to retry.
    if drop_file.uploaded_at is not None:
        return drop_file

    # The Drop may have become inactive while the file was uploading.
    if (
        drop.revoked_at is not None
        or drop.expires_at <= now
        or drop.view_count >= drop.max_views
    ):
        metadata = get_file_metadata(drop_file.storage_key)

        if metadata is not None:
            delete_file(drop_file.storage_key)

        db.delete(drop_file)
        db.commit()

        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail="Drop is no longer active",
        )

    # At this point storage_key is still:
    # pending/<drop_id>/<file_id>
    metadata = get_file_metadata(drop_file.storage_key)

    if metadata is None:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail="File upload has not completed",
        )

    actual_size = metadata["ContentLength"]
    actual_content_type = metadata.get("ContentType")

    # The browser's claimed size was what we reserved against quota.
    # Therefore the stored object must match it exactly.
    if actual_size != drop_file.size_bytes:
        delete_file(drop_file.storage_key)

        db.delete(drop_file)
        db.commit()

        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Uploaded file size does not match the reserved size",
        )

    if actual_content_type != drop_file.content_type:
        delete_file(drop_file.storage_key)

        db.delete(drop_file)
        db.commit()

        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Uploaded file type does not match the reserved type",
        )

    temporary_key = drop_file.storage_key
    final_key = f"drops/{drop.id}/{drop_file.id}"

    promote_file(
        temporary_key=temporary_key,
        final_key=final_key,
    )

    drop_file.storage_key = final_key
    drop_file.uploaded_at = now

    db.commit()
    db.refresh(drop_file)

    return drop_file

File: frontend/lib/api/drops.ts
import { bffFetch } from "./bff";

export const MIB = 1024 * 1024;
export const MAX_FILES_PER_DROP = 5;
export const MAX_FILE_SIZE = 10 * MIB;
export const MAX_DROP_FILE_SIZE = 20 * MIB;
export const ACTIVE_STORAGE_QUOTA = 30 * MIB;

export type CreateDropInput = {
  title: string;
  content: string;
  expires_at: string;
  max_views: number;
};

export type DropStatus = "active" | "expired" | "consumed" | "revoked";

export type Drop = {
  id: string;
  owner_id: string | null;
  title: string;
  content: string;
  expires_at: string;
  max_views: number;
  view_count: number;
  status: DropStatus;
  revoked_at: string | null;
  created_at: string;
  updated_at: string;
  last_accessed_at: string | null;
};

export type CreateDropResponse = Drop & { share_token: string };

export type PresignFileResponse = {
  file_id: string;
  upload_url: string;
  fields: Record<string, string>;
  expires_in: number;
};

export type DropFile = {
  id: string;
  original_name: string;
  content_type: string;
  size_bytes: number;
  uploaded_at: string;
};

export type UpdateDropInput = {
  title?: string;
  content?: string;
  max_views?: number;
};

export type PaginatedDrops = {
  items: Drop[];
  page: number;
  page_size: number;
  total: number;
};

export type GetDropsParams = {
  page?: number;
  page_size?: number;
  search?: string;
  status?: DropStatus;
};

export type UploadProgress =
  | { stage: "creating"; totalFiles: number }
  | {
      stage: "uploading";
      file: File;
      fileIndex: number;
      completedFiles: number;
      totalFiles: number;
    }
  | { stage: "complete"; completedFiles: number; totalFiles: number };

export class DropUploadError extends Error {
  constructor(
    message: string,
    readonly drop: CreateDropResponse,
    readonly completedFiles: number,
    readonly failedFile?: File,
    options?: ErrorOptions,
  ) {
    super(message, options);
    this.name = "DropUploadError";
  }
}

export async function createDrop(
  input: CreateDropInput,
): Promise<CreateDropResponse> {
  return bffFetch<CreateDropResponse>("/api/drops", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(input),
  });
}

export async function presignDropFile(
  dropId: string,
  file: File,
): Promise<PresignFileResponse> {
  return bffFetch<PresignFileResponse>(`/api/drops/${dropId}/files/presign`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      original_name: file.name,
      content_type: file.type || "application/octet-stream",
      size_bytes: file.size,
    }),
  });
}

export async function uploadToStorage(
  file: File,
  presigned: PresignFileResponse,
): Promise<void> {
  const formData = new FormData();
  for (const [key, value] of Object.entries(presigned.fields))
    formData.append(key, value);
  formData.append("file", file);
  const response = await fetch(presigned.upload_url, {
    method: "POST",
    body: formData,
  });
  if (!response.ok)
    throw new Error("The selected file could not be uploaded to storage.");
}

export async function completeDropFile(
  dropId: string,
  fileId: string,
): Promise<DropFile> {
  return bffFetch<DropFile>(`/api/drops/${dropId}/files/${fileId}/complete`, {
    method: "POST",
  });
}

export async function uploadDropFile(
  dropId: string,
  file: File,
): Promise<DropFile> {
  const presigned = await presignDropFile(dropId, file);
  await uploadToStorage(file, presigned);
  return completeDropFile(dropId, presigned.file_id);
}

export function validateDropFiles(files: File[]) {
  if (files.length > MAX_FILES_PER_DROP)
    throw new Error("You can attach up to 5 files.");
  for (const file of files) {
    if (file.size === 0) throw new Error(`${file.name} is empty.`);
    if (file.size > MAX_FILE_SIZE)
      throw new Error(`${file.name} exceeds the 10 MiB file limit.`);
    if (file.name.length > 255)
      throw new Error(`${file.name} has a file name that is too long.`);
  }
  const totalSize = files.reduce((total, file) => total + file.size, 0);
  if (totalSize > MAX_DROP_FILE_SIZE)
    throw new Error("Attachments cannot exceed 20 MiB in total.");
}

export type CreateDropWithFilesInput = {
  drop: CreateDropInput;
  files: File[];
  onProgress?: (progress: UploadProgress) => void;
};

export async function createDropWithFiles({
  drop: dropInput,
  files,
  onProgress,
}: CreateDropWithFilesInput): Promise<CreateDropResponse> {
  validateDropFiles(files);
  onProgress?.({ stage: "creating", totalFiles: files.length });
  const drop = await createDrop(dropInput);

  for (let index = 0; index < files.length; index += 1) {
    const file = files[index];
    onProgress?.({
      stage: "uploading",
      file,
      fileIndex: index,
      completedFiles: index,
      totalFiles: files.length,
    });
    try {
      await uploadDropFile(drop.id, file);
    } catch (cause) {
      throw new DropUploadError(
        `The Drop was created, but ${file.name} could not be uploaded.`,
        drop,
        index,
        file,
        { cause },
      );
    }
  }

  onProgress?.({
    stage: "complete",
    completedFiles: files.length,
    totalFiles: files.length,
  });
  return drop;
}

export async function getDrops(
  params: GetDropsParams = {},
): Promise<PaginatedDrops> {
  const searchParams = new URLSearchParams();
  if (params.page !== undefined) searchParams.set("page", String(params.page));
  if (params.page_size !== undefined)
    searchParams.set("page_size", String(params.page_size));
  if (params.search) searchParams.set("search", params.search);
  if (params.status) searchParams.set("status", params.status);
  const query = searchParams.toString();
  return bffFetch<PaginatedDrops>(`/api/drops${query ? `?${query}` : ""}`);
}

export async function getDrop(dropId: string): Promise<Drop> {
  return bffFetch<Drop>(`/api/drops/${dropId}`);
}

export async function getDropFiles(dropId: string): Promise<DropFile[]> {
  return bffFetch<DropFile[]>(`/api/drops/${dropId}/files`);
}

export async function getDropShareToken(dropId: string): Promise<string> {
  const response = await bffFetch<{ share_token: string }>(
    `/api/drops/${dropId}/share-token`,
  );
  return response.share_token;
}

export async function updateDrop(
  dropId: string,
  input: UpdateDropInput,
): Promise<Drop> {
  return bffFetch<Drop>(`/api/drops/${dropId}`, {
    method: "PATCH",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(input),
  });
}

export async function revokeDrop(dropId: string): Promise<void> {
  return bffFetch<void>(`/api/drops/${dropId}/revoke`, { method: "POST" });
}
